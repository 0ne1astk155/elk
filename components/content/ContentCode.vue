<script setup lang="ts">
const props = defineProps<{
  code: string
  lang: string
}>()

const raw = computed(() => decodeURIComponent(props.code).replace(/&#39;/g, '\''))

const langMap: Record<string, string> = {
  js: 'javascript',
  ts: 'typescript',
  vue: 'html',
}

const hightlighted = computed(() => {
  return highlightCode(raw.value, langMap[props.lang] || props.lang as any)
})
</script>

<template>
  <pre class="code-block" v-html="hightlighted" />
</template>
