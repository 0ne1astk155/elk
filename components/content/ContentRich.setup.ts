import type { Emoji } from 'masto'

defineOptions({
  name: 'ContentRich',
})

const {
  content,
  emojis,
  markdown = true,
} = defineProps<{
  content: string
  emojis?: Emoji[]
  markdown?: boolean
}>()

const emojisObject = useEmojisFallback(() => emojis)

export default () => h(
  'span',
  { class: 'content-rich', dir: 'auto' },
  contentToVNode(content, {
    emojis: emojisObject.value,
    markdown,
  }),
)
