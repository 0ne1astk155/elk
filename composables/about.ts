export interface Team {
  github: string
  display: string
  twitter: string
  mastodon: string
}

export const teams: Team[] = [
  {
    github: 'antfu',
    display: 'Anthony Fu',
    twitter: 'antfu7',
    mastodon: 'antfu@mas.to',
  },
  {
    github: 'patak-dev',
    display: 'Patak',
    twitter: 'patak_dev',
    mastodon: 'patak@webtoo.ls',
  },
  {
    github: 'danielroe',
    display: 'Daniel Roe',
    twitter: 'danielcroe',
    mastodon: 'daniel@roe.dev',
  },
  {
    github: 'sxzz',
    display: 'sxzz',
    twitter: 'sanxiaozhizi',
    mastodon: 'sxzz@mas.to',
  },
].sort(() => Math.random() - 0.5)
