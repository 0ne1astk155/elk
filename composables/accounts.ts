import { login as loginMasto } from 'masto'
import type { UserLogin } from '~/types'
import { DEFAULT_SERVER } from '~/constants'

const accounts = useLocalStorage<UserLogin[]>('nuxtodon-accounts', [], { deep: true })
const currentId = useLocalStorage<string>('nuxtodon-current-user', '')

export const currentUser = computed<UserLogin | undefined>(() => {
  let user: UserLogin | undefined
  if (currentId.value) {
    user = accounts.value.find(user => user.account?.id === currentId.value)
    if (user)
      return user
  }
  // Fallback to the first account
  return accounts.value[0]
})

export const currentServer = computed<string>(() => currentUser.value?.server || DEFAULT_SERVER)

export const useAccounts = () => accounts

export async function loginTo(user: UserLogin) {
  const existing = accounts.value.findIndex(u => u.server === user.server && u.token === user.token)
  if (existing !== -1) {
    if (currentId.value === accounts.value[existing].account?.id)
      return null
    currentId.value = user.account?.id
    return true
  }

  const masto = await loginMasto({
    url: `https://${user.server}`,
    accessToken: user.token,
  })
  const me = await masto.accounts.verifyCredentials()
  user.account = me

  accounts.value.push(user)
  currentId.value = me.id
  return true
}
