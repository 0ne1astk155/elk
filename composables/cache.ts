import LRU from 'lru-cache'
import type { Account, Status } from 'masto'

const cache = new LRU<string, any>({
  max: 1000,
})

if (process.dev)
  // eslint-disable-next-line no-console
  console.log({ cache })

export function setCached(key: string, value: any) {
  cache.set(key, value)
}

export function fetchStatus(id: string) {
  const key = `status:${id}`
  const cached = cache.get(key)
  if (cached)
    return cached
  const promise = masto.statuses.fetch(id)
    .then((status) => {
      cacheStatus(status)
      return status
    })
  cache.set(key, promise)
  return promise
}

export function fetchAccount(id: string) {
  const key = `account:${id}`
  const cached = cache.get(key)
  if (cached)
    return cached
  const promise = masto.accounts.fetch(id)
    .then((account) => {
      cacheAccount(account)
      return account
    })
  cache.set(key, promise)
  return promise
}

export function fetchAccountByName(acct: string) {
  const key = `account:${acct}`
  const cached = cache.get(key)
  if (cached)
    return cached
  const account = masto.accounts.fetch(acct)
    .then((r) => {
      cacheAccount(r)
      return r
    })
  cache.set(key, account)
  return account
}

export function cacheStatus(status: Status) {
  setCached(`status:${status.id}`, status)
}

export function cacheAccount(account: Account) {
  setCached(`account:${account.id}`, account)
  setCached(`account:${account.acct}`, account)
}
