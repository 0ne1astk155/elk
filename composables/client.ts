import { Client } from '~/api-client'

const client = new Client({
  host: 'https://mas.to',
})

export function useClient() {
  return client
}
