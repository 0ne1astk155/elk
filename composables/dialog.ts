import { STORAGE_KEY_FIRST_VISIT, STORAGE_KEY_ZEN_MODE } from '~/constants'

export const isFirstVisit = useLocalStorage(STORAGE_KEY_FIRST_VISIT, true)
export const isZenmode = useLocalStorage(STORAGE_KEY_ZEN_MODE, false)
export const toggleZenmode = useToggle(isZenmode)

export const isUserSwitcherOpen = ref(false)
export const isSigninDialogOpen = ref(false)
export const isPreviewHelpOpen = ref(isFirstVisit.value)

export function openUserSwitcher() {
  isUserSwitcherOpen.value = true
}

export function openSigninDialog() {
  isSigninDialogOpen.value = true
  isUserSwitcherOpen.value = false
}

export function openPreviewHelp() {
  isPreviewHelpOpen.value = true
}

if (isPreviewHelpOpen.value) {
  watch(isPreviewHelpOpen, () => {
    isFirstVisit.value = false
  })
}
