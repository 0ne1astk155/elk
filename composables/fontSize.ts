import type { InjectionKey, Ref } from 'vue'
import { STORAGE_KEY_FONT_SIZE } from '~/constants'

export type FontSize = 'xs' | 'sm' | 'md' | 'lg' | 'xl'
export const InjectionKeyFontSize = Symbol('fontSize') as InjectionKey<Ref<FontSize>>

export function setFontSize(size: FontSize) {
  inject(InjectionKeyFontSize)!.value = size
}

export function getFontSize() {
  return inject(InjectionKeyFontSize)!
}

export const fontSizeMap = {
  xs: '13px',
  sm: '14px',
  md: '15px',
  lg: '16px',
  xl: '17px',
}

export async function setupFontSize() {
  const fontSize = useCookie<FontSize>(STORAGE_KEY_FONT_SIZE, { default: () => 'md' })
  getCurrentInstance()?.appContext.app.provide(InjectionKeyFontSize, fontSize)

  if (!process.server) {
    watchEffect(() => {
      document.documentElement.style.setProperty('--font-size', fontSizeMap[fontSize.value || 'md'])
    })
  }
  else {
    useHead({
      style: [
        {
          innerHTML: `:root { --font-size: ${fontSizeMap[fontSize.value || 'md']}; }`,
        },
      ],
    })
  }
}
