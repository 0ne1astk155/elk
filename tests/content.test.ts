import type { Emoji } from 'masto'
import { describe, expect, it } from 'vitest'
import { renderToString } from 'vue/server-renderer'
import { contentToVNode } from '~/composables/content'

async function render(content: string, emojis?: Record<string, Emoji>) {
  const vnode = contentToVNode(content, emojis)
  const html = (await renderToString(vnode))
    .replace(/<!--[\[\]]-->/g, '')
  return {
    vnode,
    html,
  }
}

describe('rich-content', () => {
  it('plain', async () => {
    const { html } = await render('Hello\nWorld')
    expect(html).toMatchSnapshot()
  })
})
