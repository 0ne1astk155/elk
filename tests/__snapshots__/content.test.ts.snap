// Vitest Snapshot v1

exports[`rich-content > code frame 1`] = `
"<p>Testing code block</p>
<pre lang=\\"ts\\">
import { useMouse, usePreferredDark } from &#39;@vueuse/core&#39;
// tracks mouse position
const { x, y } = useMouse()
// is the user prefers dark theme
const isDark = usePreferredDark()</pre
>
<p></p>
"
`;

exports[`rich-content > custom emoji 1`] = `
"Daniel Roe
<img
  src=\\"https://media.mas.to/masto-public/cache/custom_emojis/images/000/288/667/original/c96ba3cb0e0e1eac.png\\"
  alt=\\"nuxt\\"
  class=\\"custom-emoji\\"
/>
"
`;

exports[`rich-content > empty 1`] = `""`;

exports[`rich-content > link + mention 1`] = `
"<p>
  Happy ðŸ¤— weâ€™re now using
  <span class=\\"h-card\\"
    ><a
      class=\\"u-url mention\\"
      rel=\\"nofollow noopener noreferrer\\"
      to=\\"/@vitest@mas.to\\"
    ></a
  ></span>
  (migrated from chai+mocha)
  <a
    href=\\"https://github.com/ayoayco/astro-reactive-library/pull/203\\"
    rel=\\"nofollow noopener noreferrer\\"
    target=\\"_blank\\"
    ><span class=\\"invisible\\">https://</span
    ><span class=\\"ellipsis\\">github.com/ayoayco/astro-react</span
    ><span class=\\"invisible\\">ive-library/pull/203</span></a
  >
</p>
"
`;
